src/transformers/models/audio_spectrogram_transformer/feature_extraction_audio_spectrogram_transformer.py

```diff
@@ -230,7 +247,39 @@ def __call__(
         if self.do_normalize:
             padded_inputs["input_values"] = [self.normalize(feature) for feature in input_values]
 
+        if add_noise:
+            padded_inputs["input_values"] = [self.add_noise(feature) for feature in input_values]
+
         if return_tensors is not None:
             padded_inputs = padded_inputs.convert_to_tensors(return_tensors)
 
         return padded_inputs
+
+    # from https://github.com/YuanGongND/ast with minor changes
+    def mask(self, fbank, freqm, timem):
+        if freqm == 0 and timem == 0:
+            return fbank
+        if not is_speech_available():
+            raise ImportError("Torchaudio is required to use frequency and/or time masking.")
+        if not isinstance(fbank, torch.Tensor):
+            fbank = torch.tensor(fbank)
+        fbank = torch.transpose(fbank, 0, 1)
+        # this is just to satisfy new torchaudio version, which only accept [1, freq, time]
+        fbank = fbank.unsqueeze(0)
+        if freqm != 0:
+            freqm = FrequencyMasking(freqm)
+            fbank = freqm(fbank)
+        if timem != 0:
+            timem = TimeMasking(timem)
+            fbank = timem(fbank)
+        # squeeze it back, it is just a trick to satisfy new torchaudio version
+        fbank = fbank.squeeze(0)
+        fbank = torch.transpose(fbank, 0, 1)
+        fbank = fbank.numpy()
+        return fbank
+
+    # from https://github.com/YuanGongND/ast but converted into numpy
+    def add_noise(self, fbank: np.ndarray) -> np.ndarray:
+        fbank += np.random.rand(fbank.shape[0], fbank.shape[1]) * (np.random.rand() / 10)
+        fbank = np.roll(fbank, np.random.randint(-10, 10), axis=0)
```

Comment:
Why do we use roll and randint here ? Out of curiosity:

Reply:
To add temporal shifting. I split `add_noise` into `add_noise` and `add_temporal_shift` to make it more readable as suggestion in one of your other comments.